src/Data/Filters/UserSiteMatch.php

```diff
@@ -0,0 +1,61 @@
+<?php
+/**
+ * This file provides an implementation of the UserSiteMatch filter.
+ *
+ * PHP Version 7
+ *
+ * @category   Data
+ * @package    Main
+ * @subpackage Data
+ * @author     Dave MacFarlane <david.macfarlane2@mcgill.ca>
+ * @license    http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link       https://www.github.com/aces/Loris/
+ */
+namespace LORIS\Data\Filters;
+
+/**
+ * UserSiteMatch filters out data for any resource which is not part of one of the
+ * user's sites. An Instance using the UserSiteMatch filter must implement a
+ * getCenterIDs or getCenterID method which returns an integer (or array) of
+ * CenterIDs that the data belongs to. The data will be filtered out unless the
+ * User is a member of at least one site that the resource instance is a member of.
+ *
+ * @category   Data
+ * @package    Main
+ * @subpackage Data
+ * @author     Dave MacFarlane <david.macfarlane2@mcgill.ca>
+ * @license    http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link       https://www.github.com/aces/Loris/
+ */
+class UserSiteMatch implements \LORIS\Data\Filter
+{
+    /**
+     * Implements the \LORIS\Data\Filter interface
+     *
+     * @param \User                $user     The user that the data is being
+     *                                       filtered for.
+     * @param \LORIS\Data\Instance $resource The data being filtered.
+     *
+     * @return bool true if the user has a site in common with the data
+     */
+    public function filter(\User $user, \Loris\Data\Instance $resource) : bool
+    {
+        if (method_exists($resource, 'getCenterIDs')) {
+            // If the Resource belongs to multiple CenterIDs, the user can
+            // access the data if the user is part of any of thos centers.
+            $resourceSites = $resource->getCenterIDs();
+            foreach ($resourceSites as $site) {
+                if ($user->hasCenter($resourceSite)) {
+                       return true;
+                }
+            }
+            return false;
+        } else if (method_exists($resource, 'getCenterID')) {
+            $resourceSite = $resource->getCenterID();
```

Comment:
Valid point.
What about starting to move away from IDs and work whit Sites instead?
The ressource that is UserSiteFilterable (see the interface here) should implement getSites that return an array of Site(s).  
There will be a check for `instanceof  UserSiteFilterable` then the loop to check `$user->hasCenter($resourceSite->getID)` but then we are back to IDs... 
Later User should expose a getSites function and Site should should be comparable

@driusan Any thoughts? Objections? Insults?

Reply:
If the resource was a member of multiple sites why would it have a model which has a getCenterID method and no getCenterIDs method?

I generally agrees that we should use getSites or similar instead, but none of the models that currently exist work that way right now, and IIRC the existing Site object is costly to construct

Reply:
Site is not that big. The problem is more about the full mutability of the object provided by the `select` method. 

To answer 
```If the resource was a member of multiple sites why would it have a model which has a getCenterID method and no getCenterIDs method?``` 
I think getSites() is enough.  Which goes with what Melanie said.

Reply:
After discussion with driusan, we aggreed on the current Site class not being usable as a per row model. 
getCenterIDs would be sufficient to all cases but it make no sens for something that will only ever have a single site. Since the is no  UserSiteFilterable interface driusan want this to be as flexible as it can get; hence the two methods.
I will aprouve this PR and let it to @PapillonMcGill to argue :)

Reply:
Like I said, I generally agree that getSites would be better than getCenterIDs, but the existing Site object makes (at least) one extra unnecessary query in the constructor, making it costly to require that DataInstances return Site objects rather than IDs (at least until we have a data model for Sites that doesn't have the overhead of the current one.)

I also think that this generic filter should work with as many types of data models as it can. I think it's reasonable to assume that if someone implemented a "getCenterIDs" method on their object, the resource is part of multiple centers, and if they implemented a "getCenterID" it's part of a single center and if they're implementing a model which has "getCenterID" but is part of multiple centers their model is seriously flawed/buggy to start with.

Reply:
I agree that if the model of a module is badly designed, it is the owner's problem at the base but we (as the Loris owner or main contributing centre) will have to deal with it. I just prefer to be safe than sorry in case like that.

Would it be possible that even if a module is made to work in only one center, someone would try to use it  for multiple?

Reply:
What does "using it for multiple centers" mean in the context of a filter?

It returns a bool of either true or false at the end of the day. If the getCenterIDs (plural) returns an array of one site, it'll still work (and take precedence over the singular). The class doc for this filter specifies that you have access to the resource if you have access to *any* sites that the resource is part of, so the only way I can see that going wrong is if you implement a "getCenterIDs" which returns the wrong sites, in which case again, we're back to the developer using seriously flawed model in the first place that we can't do anything about.

Reply:
I will take your word for granted on that.